---
title: "Lab-19-Pertussis"
author: "Ethan"
date: "3/14/2023"
output: html_document
---

install.packages("datapasta")

# Investigating pertussis cases by year

CDC tracking cases of Pertussis in the US. We can get their data via web scrapping


```{r echo=FALSE}
# load in data from copy/paste by 'addins' button
cdc <- data.frame(data.frame(
                                             Year = c(1922L,1923L,1924L,
                                                      1925L,1926L,1927L,1928L,
                                                      1929L,1930L,1931L,
                                                      1932L,1933L,1934L,1935L,
                                                      1936L,1937L,1938L,1939L,
                                                      1940L,1941L,1942L,
                                                      1943L,1944L,1945L,1946L,
                                                      1947L,1948L,1949L,1950L,
                                                      1951L,1952L,1953L,1954L,
                                                      1955L,1956L,1957L,
                                                      1958L,1959L,1960L,1961L,
                                                      1962L,1963L,1964L,1965L,
                                                      1966L,1967L,1968L,
                                                      1969L,1970L,1971L,1972L,
                                                      1973L,1974L,1975L,1976L,
                                                      1977L,1978L,1979L,1980L,
                                                      1981L,1982L,1983L,
                                                      1984L,1985L,1986L,1987L,
                                                      1988L,1989L,1990L,1991L,
                                                      1992L,1993L,1994L,
                                                      1995L,1996L,1997L,1998L,
                                                      1999L,2000L,2001L,2002L,
                                                      2003L,2004L,2005L,2006L,
                                                      2007L,2008L,2009L,
                                                      2010L,2011L,2012L,2013L,
                                                      2014L,2015L,2016L,2017L,
                                                      2018L,2019L),
                    Cases = c(107473,164191,
                                                      165418,152003,202210,181411,
                                                      161799,197371,166914,
                                                      172559,215343,179135,
                                                      265269,180518,147237,
                                                      214652,227319,103188,183866,
                                                      222202,191383,191890,
                                                      109873,133792,109860,
                                                      156517,74715,69479,120718,
                                                      68687,45030,37129,
                                                      60886,62786,31732,28295,
                                                      32148,40005,14809,11468,
                                                      17749,17135,13005,6799,
                                                      7717,9718,4810,3285,
                                                      4249,3036,3287,1759,
                                                      2402,1738,1010,2177,2063,
                                                      1623,1730,1248,1895,
                                                      2463,2276,3589,4195,
                                                      2823,3450,4157,4570,2719,
                                                      4083,6586,4617,5137,
                                                      7796,6564,7405,7298,
                                                      7867,7580,9771,11647,
                                                      25827,25616,15632,10454,
                                                      13278,16858,27550,18719,
                                                      48277,28639,32971,20762,
                                                      17972,18975,15609,
                                                      18617)
                   ))
```

```{r}
# Call ggplot2 pkg
library(ggplot2)
```

Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
cdc.plot <- ggplot(cdc) +
  aes(Year, Cases) +
  geom_point() +
  geom_line() +
  labs(title = "Cases of Pertussis in US from 1920 to 2019",
       subtitle = "Data from the CDC")
cdc.plot
```


Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
cdc.plot + 
  geom_vline(xintercept = 1946, linetype = 2 ,color = "blue") +
  geom_vline(xintercept = 1996, linetype = 2 ,color = "red")
```

Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

The introduction of the aP vaccine in 1996 showed a continued "low" of cases of pertussis from the wP vaccine. But, the cases started to increase afterwards. Maybe, the aP vaccine just isn't as effective.

# CMI-PB Project
Collest data on aP_ and wP individuals and their immune response to infection/booster shots

CMI-PB returns in JSON format.

On Console:
install.packages("jsonlite")

```{r}
# Allows us to read, write and process JSON data
library(jsonlite)
```

```{r}
# Read in data
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
head(subject)
```

Q4. How may aP and wP infancy vaccinated subjects are in the dataset?
47 aP and 49 wP
```{r}
# Use table to separate the column into categories
table(subject$infancy_vac)
```

Q5. How many Male and Female subjects/patients are in the dataset?
66 Female and 30 Male
```{r}
# same logic different parameters
table(subject$biological_sex)
```

Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
# again, but with 2 parameters
table(subject$race, subject$biological_sex)
```

```{r}
# working with dates
library(lubridate)
```

Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?
i)  36.36 years
ii) 25.52 years

Get age in years for all subjects:

```{r}
age_days <- today() - ymd(subject$year_of_birth)
age_years <- time_length(age_days, "years")
subject$age <- age_years
```

Average them:
```{r}
library("dplyr")
```


```{r}
# Using dplyr
# Separate the categories first. Can use this for other things
ap.age <- filter(subject, infancy_vac == "aP")$age
wp.age <- filter(subject, infancy_vac == "wP")$age

mean(ap.age)
mean(wp.age)

```

T-test
```{r}
t.test(ap.age,wp.age)
```

Q8. Determine the age of all individuals at time of boost?
```{r}
subject$age.boost <- time_length(ymd(subject$date_of_boost) - ymd(subject$year_of_birth),"years")
subject$age.boost
```



Q9. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?
Yes. There is no overlap
```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```

# Joining multiple tables

```{r}
# New table
specimen <- read_json("http://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 
```


Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
head(specimen)
head(titer)
```

```{r}
dim(subject)
dim(specimen)

meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
head(abdata)
```
Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?
IgE  IgG IgG1 IgG2 IgG3 IgG4 
6698 1413 6141 6141 6141 6141 
```{r}
table(abdata$isotype)
```

Q12. What do you notice about the number of visit 8 specimens compared to other visits?
 1    2    3    4    5    6    7    8 
5795 4640 4640 4640 4640 4320 3920   80
It's way lower. It's still ongoing. Because the data is not complete, we shouldn't use it.
```{r}
table(abdata$visit)
```

Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

```{r}
ggplot(ig1) +
  aes(MFI,antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2)
```
Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?
FIM2/3 shows change over multiple visits. 

```{r}
ggplot(ig1) +
  aes(MFI, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).

```{r}
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
    geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "Measles Antigen per Visit (aP red, wP teal)")
```

```{r}
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
    geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "FIM2/3 Antigen per Visit (aP red, wP teal)")
```

16. What do you notice about these two antigens time course and the FIM2/3 data in particular?
Over time, measles antigen stays relatively low. FIM2/3 increases per visit.

Q17. Do you see any clear difference in aP vs. wP responses?
At least in FIM2/3, aP vaccines show higher antigen response through multiple visits. However in the measles data, the aP and wP show similar responses.

# Obtaining CMI-PB RNASeq Data

```{r}
# Read in file
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"
rna <- read_json(url, simplifyVector = TRUE)

#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).
```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

Q19.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?
Maxes out at visit 4

Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?
Nope. It doesn't waver/decrease afterwards. 

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```


