---
title: "Lab-19-Pertussis"
author: "Ethan"
date: "3/14/2023"
output: pdf_document
---

install.packages("datapasta")

# Investigating pertussis cases by year

CDC tracking cases of Pertussis in the US. We can get their data via web scrapping





```r
# Call ggplot2 pkg
library(ggplot2)
```

Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.


```r
cdc.plot <- ggplot(cdc) +
  aes(Year, Cases) +
  geom_point() +
  geom_line() +
  labs(title = "Cases of Pertussis in US from 1920 to 2019",
       subtitle = "Data from the CDC")
cdc.plot
```

![](lab-19-143_files/figure-latex/unnamed-chunk-3-1.pdf)<!-- --> 


Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?


```r
cdc.plot + 
  geom_vline(xintercept = 1946, linetype = 2 ,color = "blue") +
  geom_vline(xintercept = 1996, linetype = 2 ,color = "red")
```

![](lab-19-143_files/figure-latex/unnamed-chunk-4-1.pdf)<!-- --> 

Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

The introduction of the aP vaccine in 1996 showed a continued "low" of cases of pertussis from the wP vaccine. But, the cases started to increase afterwards. Maybe, the aP vaccine just isn't as effective.

# CMI-PB Project
Collest data on aP_ and wP individuals and their immune response to infection/booster shots

CMI-PB returns in JSON format.

On Console:
install.packages("jsonlite")


```r
# Allows us to read, write and process JSON data
library(jsonlite)
```


```r
# Read in data
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
head(subject)
```

```
##   subject_id infancy_vac biological_sex              ethnicity  race
## 1          1          wP         Female Not Hispanic or Latino White
## 2          2          wP         Female Not Hispanic or Latino White
## 3          3          wP         Female                Unknown White
## 4          4          wP           Male Not Hispanic or Latino Asian
## 5          5          wP           Male Not Hispanic or Latino Asian
## 6          6          wP         Female Not Hispanic or Latino White
##   year_of_birth date_of_boost      dataset
## 1    1986-01-01    2016-09-12 2020_dataset
## 2    1968-01-01    2019-01-28 2020_dataset
## 3    1983-01-01    2016-10-10 2020_dataset
## 4    1988-01-01    2016-08-29 2020_dataset
## 5    1991-01-01    2016-08-29 2020_dataset
## 6    1988-01-01    2016-10-10 2020_dataset
```

Q4. How may aP and wP infancy vaccinated subjects are in the dataset?
47 aP and 49 wP

```r
# Use table to separate the column into categories
table(subject$infancy_vac)
```

```
## 
## aP wP 
## 47 49
```

Q5. How many Male and Female subjects/patients are in the dataset?
66 Female and 30 Male

```r
# same logic different parameters
table(subject$biological_sex)
```

```
## 
## Female   Male 
##     66     30
```

Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?


```r
# again, but with 2 parameters
table(subject$race, subject$biological_sex)
```

```
##                                            
##                                             Female Male
##   American Indian/Alaska Native                  0    1
##   Asian                                         18    9
##   Black or African American                      2    0
##   More Than One Race                             8    2
##   Native Hawaiian or Other Pacific Islander      1    1
##   Unknown or Not Reported                       10    4
##   White                                         27   13
```


```r
# working with dates
library(lubridate)
```

```
## 
## Attaching package: 'lubridate'
```

```
## The following objects are masked from 'package:base':
## 
##     date, intersect, setdiff, union
```

Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?
i)  36.36 years
ii) 25.52 years

Get age in years for all subjects:


```r
age_days <- today() - ymd(subject$year_of_birth)
age_years <- time_length(age_days, "years")
subject$age <- age_years
```

Average them:

```r
library("dplyr")
```

```
## 
## Attaching package: 'dplyr'
```

```
## The following objects are masked from 'package:stats':
## 
##     filter, lag
```

```
## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union
```



```r
# Using dplyr
# Separate the categories first. Can use this for other things
ap.age <- filter(subject, infancy_vac == "aP")$age
wp.age <- filter(subject, infancy_vac == "wP")$age

mean(ap.age)
```

```
## [1] 25.5156
```

```r
mean(wp.age)
```

```
## [1] 36.36006
```

T-test

```r
t.test(ap.age,wp.age)
```

```
## 
## 	Welch Two Sample t-test
## 
## data:  ap.age and wp.age
## t = -12.092, df = 51.082, p-value < 2.2e-16
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -12.644857  -9.044045
## sample estimates:
## mean of x mean of y 
##  25.51560  36.36006
```

Q8. Determine the age of all individuals at time of boost?

```r
subject$age.boost <- time_length(ymd(subject$date_of_boost) - ymd(subject$year_of_birth),"years")
subject$age.boost
```

```
##  [1] 30.69678 51.07461 33.77413 28.65982 25.65914 28.77481 35.84942 34.14921
##  [9] 20.56400 34.56263 30.65845 34.56263 19.56194 23.61944 27.61944 29.56331
## [17] 36.69815 19.65777 22.73511 32.26557 25.90007 23.90144 25.90007 28.91992
## [25] 42.92129 47.07461 47.07461 29.07324 21.07324 21.07324 28.15058 24.15058
## [33] 24.15058 21.14990 21.14990 31.20876 26.20671 32.20808 27.20876 26.20671
## [41] 21.20739 20.26557 22.26420 19.32375 21.32238 19.32375 19.32375 22.41752
## [49] 20.41889 21.41821 19.47707 23.47707 20.47639 21.47570 19.47707 35.65777
## [57] 33.65914 31.65777 25.73580 24.70089 28.70089 33.73580 19.73443 34.73511
## [65] 19.73443 28.73648 27.73443 19.81109 26.77344 33.81246 25.77413 19.81109
## [73] 18.85010 19.81109 31.81109 22.81177 31.84942 19.84942 18.85010 18.85010
## [81] 19.90691 18.85010 20.90897 19.04449 20.04381 19.90691 19.90691 19.00616
## [89] 19.00616 20.04381 20.04381 20.07940 21.08145 20.07940 20.07940 20.07940
```



Q9. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?
Yes. There is no overlap

```r
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```

```
## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
```

![](lab-19-143_files/figure-latex/unnamed-chunk-16-1.pdf)<!-- --> 

# Joining multiple tables


```r
# New table
specimen <- read_json("http://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE) 
```


Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:


```r
head(specimen)
```

```
##   specimen_id subject_id actual_day_relative_to_boost
## 1           1          1                           -3
## 2           2          1                          736
## 3           3          1                            1
## 4           4          1                            3
## 5           5          1                            7
## 6           6          1                           11
##   planned_day_relative_to_boost specimen_type visit
## 1                             0         Blood     1
## 2                           736         Blood    10
## 3                             1         Blood     2
## 4                             3         Blood     3
## 5                             7         Blood     4
## 6                            14         Blood     5
```

```r
head(titer)
```

```
##   specimen_id isotype is_antigen_specific antigen        MFI MFI_normalised
## 1           1     IgE               FALSE   Total 1110.21154       2.493425
## 2           1     IgE               FALSE   Total 2708.91616       2.493425
## 3           1     IgG                TRUE      PT   68.56614       3.736992
## 4           1     IgG                TRUE     PRN  332.12718       2.602350
## 5           1     IgG                TRUE     FHA 1887.12263      34.050956
## 6           1     IgE                TRUE     ACT    0.10000       1.000000
##    unit lower_limit_of_detection
## 1 UG/ML                 2.096133
## 2 IU/ML                29.170000
## 3 IU/ML                 0.530000
## 4 IU/ML                 6.205949
## 5 IU/ML                 4.679535
## 6 IU/ML                 2.816431
```


```r
dim(subject)
```

```
## [1] 96 10
```

```r
dim(specimen)
```

```
## [1] 729   6
```

```r
meta <- inner_join(specimen, subject)
```

```
## Joining with `by = join_by(subject_id)`
```

```r
dim(meta)
```

```
## [1] 729  15
```

```r
head(meta)
```

```
##   specimen_id subject_id actual_day_relative_to_boost
## 1           1          1                           -3
## 2           2          1                          736
## 3           3          1                            1
## 4           4          1                            3
## 5           5          1                            7
## 6           6          1                           11
##   planned_day_relative_to_boost specimen_type visit infancy_vac biological_sex
## 1                             0         Blood     1          wP         Female
## 2                           736         Blood    10          wP         Female
## 3                             1         Blood     2          wP         Female
## 4                             3         Blood     3          wP         Female
## 5                             7         Blood     4          wP         Female
## 6                            14         Blood     5          wP         Female
##                ethnicity  race year_of_birth date_of_boost      dataset
## 1 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 2 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 3 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 4 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 5 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 6 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
##        age age.boost
## 1 37.19644  30.69678
## 2 37.19644  30.69678
## 3 37.19644  30.69678
## 4 37.19644  30.69678
## 5 37.19644  30.69678
## 6 37.19644  30.69678
```


```r
abdata <- inner_join(titer, meta)
```

```
## Joining with `by = join_by(specimen_id)`
```

```r
dim(abdata)
```

```
## [1] 32675    22
```

```r
head(abdata)
```

```
##   specimen_id isotype is_antigen_specific antigen        MFI MFI_normalised
## 1           1     IgE               FALSE   Total 1110.21154       2.493425
## 2           1     IgE               FALSE   Total 2708.91616       2.493425
## 3           1     IgG                TRUE      PT   68.56614       3.736992
## 4           1     IgG                TRUE     PRN  332.12718       2.602350
## 5           1     IgG                TRUE     FHA 1887.12263      34.050956
## 6           1     IgE                TRUE     ACT    0.10000       1.000000
##    unit lower_limit_of_detection subject_id actual_day_relative_to_boost
## 1 UG/ML                 2.096133          1                           -3
## 2 IU/ML                29.170000          1                           -3
## 3 IU/ML                 0.530000          1                           -3
## 4 IU/ML                 6.205949          1                           -3
## 5 IU/ML                 4.679535          1                           -3
## 6 IU/ML                 2.816431          1                           -3
##   planned_day_relative_to_boost specimen_type visit infancy_vac biological_sex
## 1                             0         Blood     1          wP         Female
## 2                             0         Blood     1          wP         Female
## 3                             0         Blood     1          wP         Female
## 4                             0         Blood     1          wP         Female
## 5                             0         Blood     1          wP         Female
## 6                             0         Blood     1          wP         Female
##                ethnicity  race year_of_birth date_of_boost      dataset
## 1 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 2 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 3 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 4 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 5 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 6 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
##        age age.boost
## 1 37.19644  30.69678
## 2 37.19644  30.69678
## 3 37.19644  30.69678
## 4 37.19644  30.69678
## 5 37.19644  30.69678
## 6 37.19644  30.69678
```
Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?
IgE  IgG IgG1 IgG2 IgG3 IgG4 
6698 1413 6141 6141 6141 6141 

```r
table(abdata$isotype)
```

```
## 
##  IgE  IgG IgG1 IgG2 IgG3 IgG4 
## 6698 1413 6141 6141 6141 6141
```

Q12. What do you notice about the number of visit 8 specimens compared to other visits?
 1    2    3    4    5    6    7    8 
5795 4640 4640 4640 4640 4320 3920   80
It's way lower. It's still ongoing. Because the data is not complete, we shouldn't use it.

```r
table(abdata$visit)
```

```
## 
##    1    2    3    4    5    6    7    8 
## 5795 4640 4640 4640 4640 4320 3920   80
```

Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:


```r
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

```
##   specimen_id isotype is_antigen_specific antigen        MFI MFI_normalised
## 1           1    IgG1                TRUE     ACT 274.355068      0.6928058
## 2           1    IgG1                TRUE     LOS  10.974026      2.1645083
## 3           1    IgG1                TRUE   FELD1   1.448796      0.8080941
## 4           1    IgG1                TRUE   BETV1   0.100000      1.0000000
## 5           1    IgG1                TRUE   LOLP1   0.100000      1.0000000
## 6           1    IgG1                TRUE Measles  36.277417      1.6638332
##    unit lower_limit_of_detection subject_id actual_day_relative_to_boost
## 1 IU/ML                 3.848750          1                           -3
## 2 IU/ML                 4.357917          1                           -3
## 3 IU/ML                 2.699944          1                           -3
## 4 IU/ML                 1.734784          1                           -3
## 5 IU/ML                 2.550606          1                           -3
## 6 IU/ML                 4.438966          1                           -3
##   planned_day_relative_to_boost specimen_type visit infancy_vac biological_sex
## 1                             0         Blood     1          wP         Female
## 2                             0         Blood     1          wP         Female
## 3                             0         Blood     1          wP         Female
## 4                             0         Blood     1          wP         Female
## 5                             0         Blood     1          wP         Female
## 6                             0         Blood     1          wP         Female
##                ethnicity  race year_of_birth date_of_boost      dataset
## 1 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 2 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 3 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 4 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 5 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
## 6 Not Hispanic or Latino White    1986-01-01    2016-09-12 2020_dataset
##        age age.boost
## 1 37.19644  30.69678
## 2 37.19644  30.69678
## 3 37.19644  30.69678
## 4 37.19644  30.69678
## 5 37.19644  30.69678
## 6 37.19644  30.69678
```


```r
ggplot(ig1) +
  aes(MFI,antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2)
```

![](lab-19-143_files/figure-latex/unnamed-chunk-24-1.pdf)<!-- --> 
Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?
FIM2/3 shows change over multiple visits. 


```r
ggplot(ig1) +
  aes(MFI, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

![](lab-19-143_files/figure-latex/unnamed-chunk-25-1.pdf)<!-- --> 

Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).


```r
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
    geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "Measles Antigen per Visit (aP red, wP teal)")
```

![](lab-19-143_files/figure-latex/unnamed-chunk-26-1.pdf)<!-- --> 


```r
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
    geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "FIM2/3 Antigen per Visit (aP red, wP teal)")
```

![](lab-19-143_files/figure-latex/unnamed-chunk-27-1.pdf)<!-- --> 

16. What do you notice about these two antigens time course and the FIM2/3 data in particular?
Over time, measles antigen stays relatively low. FIM2/3 increases per visit.

Q17. Do you see any clear difference in aP vs. wP responses?
At least in FIM2/3, aP vaccines show higher antigen response through multiple visits. However in the measles data, the aP and wP show similar responses.

# Obtaining CMI-PB RNASeq Data


```r
# Read in file
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"
rna <- read_json(url, simplifyVector = TRUE)

#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

```
## Joining with `by = join_by(specimen_id)`
```

Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```r
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

![](lab-19-143_files/figure-latex/unnamed-chunk-29-1.pdf)<!-- --> 

Q19.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?
Maxes out at visit 4

Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?
Nope. It doesn't waver/decrease afterwards. 


```r
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

![](lab-19-143_files/figure-latex/unnamed-chunk-30-1.pdf)<!-- --> 


```r
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() 
```

![](lab-19-143_files/figure-latex/unnamed-chunk-31-1.pdf)<!-- --> 


